src/bin/pg_autoctl/pgctl.c

```diff
@@ -835,38 +907,78 @@ pg_ctl_promote(const char *pg_ctl, const char *pgdata)
-					   const char *replicationUsername, const char *replicationPassword,
+pg_setup_standby_mode(uint32_t pg_control_version,
+					  const char *configFilePath,
+					  const char *pgdata,
+					  ReplicationSource *replicationSource)
+{
+	NodeAddress *primaryNode = &(replicationSource->primaryNode);
+	char primaryConnInfo[MAXCONNINFO] = { 0 };
+
+	/* we ignore the length returned by prepare_primary_conninfo... */
+	(void) prepare_primary_conninfo(primaryConnInfo,
+									MAXCONNINFO,
+									primaryNode->host,
+									primaryNode->port,
+									replicationSource->userName,
+									replicationSource->password);
+
+	if (pg_control_version < 1200)
+	{
+		/*
+		 * Before Postgres 12 we used to place recovery configuration in a
+		 * specific file recovery.conf, located alongside postgresql.conf.
+		 * Controling whether the server would start in PITR or standby mode
+		 * was controlled by a setting in the recovery.conf file.
+		 */
+		return pg_write_recovery_conf(pgdata,
+									  primaryConnInfo,
+									  replicationSource->slotName);
+	}
+	else
+	{
+		/*
+		 * Starting in Postgres 12 we need to add our recovery configuration to
+		 * the main postgresql.conf file and create an empty standby.signal
+		 * file to trigger starting the server in stanby mode.
+		 */
+		return pg_write_standby_signal(configFilePath,
+									   pgdata,
+									   primaryConnInfo,
+									   replicationSource->slotName);
+	}
+}
+
+
+/*
+ * pg_write_recovery_conf writes a recovery.conf file to a postgres data
+ * directory with the given primary connection info and replication slot name.
+ */
+static bool
+pg_write_recovery_conf(const char *pgdata,
+					   const char *primaryConnInfo,
 					   const char *replicationSlotName)
 {
 	char recoveryConf[BUFSIZE] = { 0 };
 	char *recoveryConfEnd = recoveryConf;
-	char primaryConnInfo[MAXCONNINFO] = { 0 };
-	char *connInfoEnd = primaryConnInfo;
 	char recoveryConfPath[MAXPGPATH];
 	int recoveryConfLength = 0;
 
-	connInfoEnd += make_conninfo_field_str(connInfoEnd, "host", primaryHost);
-	connInfoEnd += make_conninfo_field_int(connInfoEnd, "port", primaryPort);
-	connInfoEnd += make_conninfo_field_str(connInfoEnd, "user", replicationUsername);
-
-	if (replicationPassword != NULL)
-	{
-		connInfoEnd += make_conninfo_field_str(connInfoEnd, "password",
-											   replicationPassword);
-	}
+	log_trace("pg_write_recovery_conf");
 
 	/* build the contents of recovery.conf */
 	recoveryConfEnd += sprintf(recoveryConfEnd, "standby_mode = 'on'");
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nprimary_conninfo = ");
-	recoveryConfEnd += escape_recovery_conf_string(recoveryConfEnd, primaryConnInfo);
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nprimary_slot_name = '%s'",
-							   replicationSlotName);
-	recoveryConfEnd += sprintf(recoveryConfEnd, "\nrecovery_target_timeline = 'latest'");
+	recoveryConfEnd += sprintf(recoveryConfEnd,
```

Comment:
could you consider using `StringInfo`, `appendStringInfo()` instead ? I know it was like this before. It is a good opportunity to fix that.

Reply:
See https://github.com/citusdata/pg_auto_failover/issues/54